modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/indexreader/IgniteIndexReader.java

```diff
@@ -1219,6 +1237,42 @@ private TreeNode getTreeNode(long pageId, TreeTraverseContext nodeCtx) {
         }
     }
 
+    /**
+     * @param nodeCtx tree traverse context.
+     * @param pageContent inner page content.
+     * @param io inner page io.
+     * @throws IgniteCheckedException
+     */
+    private void checkTriangleInvariant(TreeTraverseContext nodeCtx, PageContent pageContent,
+        PageIO io) throws IgniteCheckedException {
+        if (io instanceof BPlusInnerIO && pageContent.linkedPageIds != null) {
+            long rightChildPageId = -1;
+
+            // linkedPageIds are populated in InnerPageIOProcessor.getContent
+            for (Long linkedPageId: pageContent.linkedPageIds) {
+                triangleChecksNum++;
+
+                if (rightChildPageId != -1 && linkedPageId != rightChildPageId)
+                    triangleErrorsCounter++;
+
+                final ByteBuffer innerBuf = allocateBuffer(pageSize);
```

Comment:
I think it's too wasteful to allocate and deallocate buffer for every linked page, please reuse it where possible

Reply:
Fixed